src/main/java/crazypants/enderio/machine/drain/FluidHelper.java

```diff
@@ -0,0 +1,328 @@
+package crazypants.enderio.machine.drain;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockLiquid;
+import net.minecraft.world.World;
+import net.minecraftforge.common.util.ForgeDirection;
+import net.minecraftforge.fluids.BlockFluidClassic;
+import net.minecraftforge.fluids.BlockFluidFinite;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidRegistry;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.IFluidBlock;
+import crazypants.util.BlockCoord;
+
+public final class FluidHelper {
+
+  private final World world;
+  private final FluidStack stack;
+  private final Fluid fluid;
+  private final Block block;
+  private final ForgeDirection densityDir;
+  private final FType type;
+  private final BlockCoord startbc;
+  private IDrainingCallback hook;
+
+  /*
+   * Set this to block removed water from forming infinite pools. The block expires when the given te is gc()ed.
+   */
+  public void setDrainingCallback(IDrainingCallback hook) {
+    this.hook = hook;
+  }
+
+  private FluidHelper(World world, FluidStack stack, BlockCoord startbc) throws Exception {
+    this.world = world;
+    this.stack = stack;
+    this.fluid = stack.getFluid();
+    this.block = fluid.getBlock();
+    this.densityDir = fluid.getDensity() > 0 ? ForgeDirection.DOWN : ForgeDirection.UP;
+    if (block instanceof BlockFluidClassic) {
+      this.type = FType.CLASSIC;
+    } else if (block instanceof BlockFluidFinite) {
+      this.type = FType.FINITE;
+    } else if (block instanceof BlockLiquid) {
+      this.type = FType.VANILLA;
+    } else {
+      throw new Exception();
+    }
+    this.startbc = startbc;
+  }
+
+  public static boolean isSourceBlock(World world, BlockCoord bc) {
+    Block block = bc.getBlock(world);
+    if (block instanceof BlockFluidClassic) {
+      return ((BlockFluidClassic) block).isSourceBlock(world, bc.x, bc.y, bc.z);
+    } else if (block instanceof BlockFluidFinite) {
+      return ((BlockFluidFinite)block).canDrain(world, bc.x, bc.y, bc.z);
+    } else if (block instanceof BlockLiquid) {
+      return world.getBlockMetadata(bc.x, bc.y, bc.z) == 0;
+    } else {
+      return false;
+    }
+  }
+
+  private static final ForgeDirection[] DIRECTIONS_INIT = {ForgeDirection.UP, ForgeDirection.NORTH, ForgeDirection.SOUTH, ForgeDirection.WEST, ForgeDirection.EAST};
+  
+  public static FluidHelper getInstance(World world, BlockCoord bc) {
+    for (ForgeDirection forgeDirection : DIRECTIONS_INIT) {
+      BlockCoord direction = bc.getLocation(forgeDirection);
+      if (isSourceBlock(world, direction)) {
```

Comment:
Sorry, I'm not 100% sure which of the lines that Github shows me you commented on.

The isSourceBlock() certainly is no simple null check; code is just above.